src/graph/matrix/delta_matrix.rs

```diff
@@ -0,0 +1,531 @@
+
+        m
+    }
+
+    pub fn pending(&self) -> bool {
+        if self
+            .transposed
+            .as_ref()
+            .map(|t| t.pending())
+            .unwrap_or_default()
+        {
+            return true;
+        }
+
+        self.matrix.pending() || self.delta_plus.pending() || self.delta_minus.pending()
+    }
+
+    pub fn wait(
+        &mut self,
+        force_sync: bool,
+    ) {
+        if let Some(t) = self.transposed.as_mut() {
+            t.wait(force_sync);
+        }
+
+        let mut delta_max_pending_changes = MaybeUninit::uninit();
+        unsafe {
+            Config_Option_get(
+                ConfigOptionField::DELTA_MAX_PENDING_CHANGES,
+                delta_max_pending_changes.as_mut_ptr(),
+            )
+        };
+        let delta_max_pending_changes = unsafe { delta_max_pending_changes.assume_init() };
+
+        self.sync(force_sync, delta_max_pending_changes);
+        self.set_dirty(false);
+    }
+
+    fn sync(
+        &mut self,
+        force_sync: bool,
+        delta_max_pending_changes: u64,
+    ) {
+        if force_sync {
+            self.sync_deletions();
+            self.sync_additions();
+        } else {
+            if self.delta_minus.nvals() > delta_max_pending_changes {
+                self.sync_deletions();
+            }
+
+            if self.delta_plus.nvals() > delta_max_pending_changes {
+                self.sync_additions();
+            }
+        }
+
+        self.matrix.wait();
+        self.delta_plus.wait();
+        self.delta_minus.wait();
+    }
+
+    fn sync_deletions(&mut self) {
+        self.matrix
+            .transpose(Some(&self.delta_minus), null_mut(), None, unsafe {
+                GrB_DESC_RSCT0
+            });
+        self.delta_minus.clear();
+    }
+
+    fn sync_additions(&mut self) {
+        unsafe {
+            self.matrix.assign(
+                self.delta_plus.grb_matrix_ref(),
+                self.delta_plus.grb_matrix_ref(),
+                GrB_ALL,
+                self.nrows(),
+                GrB_ALL,
+                self.ncols(),
+                GrB_DESC_S,
+            );
+        }
+        self.delta_plus.clear();
+    }
+
+    pub fn extract_row(
+        &self,
+        v: GrB_Vector,
+        i: u64,
+    ) {
+        unsafe {
+            let mut vmask = MaybeUninit::uninit();
+            GrB_Vector_new(vmask.as_mut_ptr(), GrB_BOOL, self.ncols());
+            let mask = vmask.assume_init();
+
+            self.delta_minus
+                .extract(mask, null_mut(), null_mut(), i, GrB_DESC_T0);
+            self.matrix.extract(v, mask, null_mut(), i, GrB_DESC_SCT0);
+            self.delta_plus
+                .extract(v, mask, GxB_LOR_BOOL, i, GrB_DESC_SCT0);
+            GrB_Vector_free(vmask.as_mut_ptr());
+        }
+    }
+
+    pub fn synchronize(
+        &mut self,
+        nrows: u64,
+        ncols: u64,
+    ) {
+        if !(self.nrows() < nrows || self.ncols() < ncols) && !self.dirty() {
```

Comment:
Technically, the !self.dirty can just be another || inside the parenthesis?

Reply:
If !self.dirty() is likely, it can even be moved up in the order? maybe gain another yoctosecond of performance there :laughing:

Reply:
this is not

Reply:
it is not clear which is likely inserted the dirty to the or